### PR TITLE
Implement basic Component discovery

### PR DESCRIPTION
**Pull request by rynowak**
This is the basics of component discovery along with some tests.

The next set of changes will integrate it into the compilation process.

### CODE REVIEW

File: test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            // Act
+            provider.Execute(context);
+
+            // Assert
+            var components = ExcludeBuiltInComponents(context);
+            var component = Assert.Single(components);
+
+            // These are features Components don't use. Verifying them once here and
+            // then ignoring them.
+            Assert.Empty(component.AllowedChildTags);
```
Comment (SteveSandersonMS) [inline test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs]: Does this being empty mean that all child markup is allowed, or that none is? We'll want to allow any markup inside a component, e.g., for cases like

    <MyDialog>
        Some arbitrary HTML goes here, possibly containing more component tag helpers
    </MyDialog>

File: test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            Assert.Equal("System.Boolean", attribute.TypeName);
+
+            Assert.False(attribute.HasIndexer);
+            Assert.True(attribute.IsBooleanProperty);
+            Assert.False(attribute.IsEnum);
+            Assert.False(attribute.IsStringProperty);
+        }
+
+        [Fact] // enum properties have some special intellisense behavior
+        public void Excecute_EnumProperty_CreatesDescriptor()
```
Comment (SteveSandersonMS) [inline test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs]: Not necessary to fix, but typo in test name

**Comment by SteveSandersonMS** [comment]
@rynowak Looks good!

File: test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs (hunk 1 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Blazor.Razor.Extensions.Test/ComponentTagHelperDescriptorProviderTest.cs]: All properly nested markup is allowed

**Pull request merged**